activesupport/lib/active_support/cache/serializer_with_fallback.rb

```diff
@@ -76,8 +76,8 @@ def try_dump_bare_string(entry)
           return if !signature
 
           packed = [signature, entry.expires_at || -1.0, version&.bytesize || -1].pack(BARE_STRING_TEMPLATE)
-          packed << version if version
-          packed << value
+          packed << version.b if version
+          packed << value.b
```

Comment:
This always allocates, which the surrounding code seems to be trying fairly hard to avoid.

Appending `a*a*` to BARE_STRING_TEMPLATE might give the same result for free?

Reply:
When I originally wrote this code, I actually wanted to use `pack("a*")` for readability.  Unfortunately, `pack("a*")` is surprisingly slow compared to `<<`.  (Perhaps it does more internal type checking of its arguments?)

<details>
  <summary>Benchmark Script</summary>

  ```ruby
  # frozen_string_literal: true

  require "benchmark/ips"

  [
    "Needs more umlaut!" * 50,
    "Needs more ümlaut!" * 50,
  ].each do |string|
    puts
    puts "=== ascii_only?: #{string.ascii_only?}"
    puts

    Benchmark.ips do |x|
      x.report('pack("a*")') do
        packed = [0, 0, 0, string].pack("CEl<a*")
      end

      x.report("string.b") do
        packed = [0, 0, 0].pack("CEl<")
        packed << string.b
      end

      x.report("string.b ?") do
        packed = [0, 0, 0].pack("CEl<")
        packed << (string.ascii_only? ? string : string.b)
      end

      x.compare!
    end
  end
  ```
</details>

```
=== ascii_only?: true

Warming up --------------------------------------
          pack("a*")    84.382k i/100ms
            string.b    96.070k i/100ms
          string.b ?    95.627k i/100ms
Calculating -------------------------------------
          pack("a*")    842.892k (± 1.2%) i/s -      4.219M in   5.006183s
            string.b    977.256k (± 2.6%) i/s -      4.900M in   5.017213s
          string.b ?    944.848k (± 1.3%) i/s -      4.781M in   5.061278s

Comparison:
            string.b:   977255.6 i/s
          string.b ?:   944848.0 i/s - same-ish: difference falls within error
          pack("a*"):   842891.8 i/s - 1.16x  slower


=== ascii_only?: false

Warming up --------------------------------------
          pack("a*")    79.612k i/100ms
            string.b    95.322k i/100ms
          string.b ?    89.018k i/100ms
Calculating -------------------------------------
          pack("a*")    789.417k (± 0.8%) i/s -      3.981M in   5.042813s
            string.b    961.234k (± 2.6%) i/s -      4.861M in   5.061186s
          string.b ?    910.360k (± 2.6%) i/s -      4.629M in   5.088419s

Comparison:
            string.b:   961234.2 i/s
          string.b ?:   910360.1 i/s - 1.06x  slower
          pack("a*"):   789417.3 i/s - 1.22x  slower
```

It would be nice if Ruby provided a way to concatenate bytes (ignoring encoding).  Something similar was proposed in [Ruby #14975](https://bugs.ruby-lang.org/issues/14975).